Cron/ReconcileLostPayments.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+namespace Iwoca\Iwocapay\Cron;
+
+use Iwoca\Iwocapay\Model\Config;
+use Iwoca\Iwocapay\Model\IwocaClientFactory;
+use Iwoca\Iwocapay\Controller\Process\Callback;
+use Magento\Sales\Model\Order;
+use Magento\Sales\Model\OrderFactory;
+use Magento\Sales\Model\ResourceModel\Order\CollectionFactory;
+use Magento\Framework\Stdlib\DateTime\TimezoneInterface;
+use Magento\Framework\Serialize\Serializer\Json;
+use Magento\Framework\Exception\LocalizedException;
+use Psr\Log\LoggerInterface;
+use GuzzleHttp\Exception\GuzzleException;
+
+use Magento\Framework\App\ResourceConnection;
+
+/**
+ * This class attempts to reconcile payments that have dropped off early e.g. they
+ * fell out of the flow after draw-down but before redirecting back to the merchant.
+ *
+ * It also prolongs iwocaPay orders so that they aren't automatically cleared after
+ * a user set timeout. Instead, 48 hours is used.
+ */
+class ReconcileLostPayments
+{
+    /**
+     * @var OrderFactory
+     */
+    protected $orderFactory;
+    /**
+     * @var IwocaClientFactory
+     */
+    protected $iwocaClientFactory;
+    /**
+     * @var Config
+     */
+    private $config;
+    /**
+     * @var Json
+     */
+    private $jsonSerializer;
+    /**
+     * @var CollectionFactory
+     */
+    protected $orderCollectionFactory;
+    /**
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * @var TimezoneInterface
+     */
+    protected $timezone;
+
+    /**
+     * @var ResourceConnection
+     */
+    protected $resourceConnection;
+
+    public function __construct(
+        CollectionFactory  $orderCollectionFactory,
+        LoggerInterface    $logger,
+        OrderFactory       $orderFactory,
+        IwocaClientFactory $iwocaClientFactory,
+        Config             $config,
+        Json               $jsonSerializer,
+        TimezoneInterface  $timezone,
+        ResourceConnection $resourceConnection
+    )
+    {
+        $this->orderCollectionFactory = $orderCollectionFactory;
+        $this->logger = $logger;
+        $this->orderFactory = $orderFactory;
+        $this->iwocaClientFactory = $iwocaClientFactory;
+        $this->config = $config;
+        $this->jsonSerializer = $jsonSerializer;
+        $this->timezone = $timezone;
+        $this->resourceConnection = $resourceConnection;
+    }
+
+    /**
+     * Returns all orders which were completed using iwocaPay and are in the "pending payment" state
+     *
+     * @return array
+     */
+    protected function getPendingIwocaPayOrders(): array
+    {
+        $orders = [];
+        foreach ($this->orderFactory->create()->getCollection() as $order) {
+            if ($order->getPayment()->getMethod() === 'iwocapay' && $order->getStatus() === 'pending_payment') {
+                $customerId = $order->getCustomerId();
+                $orders[$customerId][] = $order;
```

Comment:
I don't recall if guest checkout allows the use of Iwoca, but for orders placed by guests the ID would be 0 I believe. If there are multiple guest orders in the collection that would mean only 1 will be updated as the array key would be the same for those orders

Reply:
Great point! Thank you @Skullsneeze. I've re-factored this function so that it's no longer customerId centric